Print tier in DumpJittedMethods.

sandreenko opened the pull request:
I found that useful when reproducing bugs that require `complus_TieredCompilation=1`.

Example:
```
Compiling    0 System.SpanHelpers::IndexOf, IL size = 902, hsh=0xda613814 for Tier-1
Compiling    1 System.Text.Unicode.Utf16Utility::GetPointerToFirstInvalidChar, IL size = 1107, hsh=0xe576e43d
Compiling    2 System.Runtime.Intrinsics.Vector128::CreateScalarUnsafe, IL size = 21, hsh=0x6746a268 for Tier-0
```

Commit:
Print tier in DumpJittedMethods.

I found that useful when reproducing bugs that require `complus_TieredCompilation=1`.

Example:
```
Compiling    0 System.SpanHelpers::IndexOf, IL size = 902, hsh=0xda613814 for Tier-1
Compiling    1 System.Text.Unicode.Utf16Utility::GetPointerToFirstInvalidChar, IL size = 1107, hsh=0xe576e43d
Compiling    2 System.Runtime.Intrinsics.Vector128::CreateScalarUnsafe, IL size = 21, hsh=0x6746a268 for Tier-0
```

sandreenko commented:
PTAL @dotnet/jit-contrib

erozenfeld commented:
The second method in the example doesn't have the tier printed. Can we figure out what tier we are compiling or have something else printed?

sandreenko commented:
> The second method in the example doesn't have the tier printed. Can we figure out what tier we are compiling or have something else printed?

I remember that I was confused by such methods as well, my understanding is that it means "normal" mode like we have with `complus_TieredCompilation=0`, for example, it happens when we compile a method with an aggressive optimization flag or recompile a crossgened method. 

@kouvel, could you please check if I am right or not?

CarolEidt commented:
This is a **great** idea!

@BruceForstall recently pointed out to me that `COMPlus_JitFunctionTrace` does much the same thing as `COMPlus_DumpJittedMethods`, so I use the former more often these days (it prints the code address, which is often convenient in the debugger). Would it make sense to add this to that dump as well? Or should we just consolidate the two?

kouvel commented:
> @kouvel, could you please check if I am right or not?

The tier flag is not set for methods that are jitted with optimizations the first time around, except for methods marked with `AggressiveOptimization`. So currently:
- Methods marked with `AggressiveOptimization` will have the tier 1 flag set (in hopes that the flag may be used to more aggressively optimize these in the future)
- Methods that contain loops are currently not tiered and the JIT removes the tier 0 flag early (see calls to `fgSwitchToOptimized()`) and will not have the tier flag set after that
- Methods that contain an explicit tail call are also not tiered and the tier 0 flag is removed similarly to above
- Rejits initiated by tiering would have the tier 1 flag set

Commit:
Fix JitFunctionTrace printing.

The opening is using decimal number, but the closing was using hex.

Commit:
Add compGetTieringName.

sandreenko commented:
> @BruceForstall recently pointed out to me that COMPlus_JitFunctionTrace does much the same thing as COMPlus_DumpJittedMethods, so I use the former more often these days (it prints the code address, which is often convenient in the debugger). Would it make sense to add this to that dump as well? Or should we just consolidate the two?

I have updated the PR to print the mode in both `JitFunctionTrace` and `DumpJittedMethods` and added `No-Tier` when both flags are unset.

BruceForstall commented on src/jit/compiler.cpp:
```diff
@@ -4272,10 +4293,14 @@ void Compiler::compFunctionTraceEnd(void* methodCodePtr, ULONG methodCodeSize, b
         }
+
+        // Note: that is incorrect if we are compiling several methods at the same time.
+        unsigned methodNumber = Compiler::jitTotalMethodCompiled - 1;
```
We should create a Compiler member that captures this once and then use that for all output, and use InterlockedIncrement to update this static.
